skbeam/core/correlation.py

```diff
@@ -842,3 +850,322 @@ def one_time_from_two_time(two_time_corr):
+            self_correlation = True
+            img2 = img1
+        else:
+            self_correlation = False
+
+        ccorrs = list()
+        rngiter = tqdm(range(self.nids))
+
+        for i in rngiter:
+            self.tmpimgs[i] *= 0
+            self.tmpimgs[i].ravel()[
+                                    self.subpxlsts[i]
+                                    ] = img1.ravel()[self.pxlsts[i]]
+            if not self_correlation:
+                self.tmpimgs2[i] *= 0
+                self.tmpimgs2[i].ravel()[
+                                        self.subpxlsts[i]
+                                        ] = img2.ravel()[self.pxlsts[i]]
+
+            # multiply by maskcorrs > 0 to ignore invalid regions
+            if self_correlation:
+                ccorr = _cross_corr(self.tmpimgs[i])*(self.maskcorrs[i] > 0)
+            else:
+                ccorr = _cross_corr(self.tmpimgs[i], self.tmpimgs2[i]) * \
+                        (self.maskcorrs[i] > 0)
+            # only select valid regions
+            w = self.pxlst_maskcorrs[i]
+
+            # now handle the normalizations
+            if 'symavg' in normalization:
+                # do symmetric averaging
+                Icorr = _cross_corr(self.tmpimgs[i] *
+                                    self.submasks[i], self.submasks[i])
+                if self_correlation:
+                    Icorr2 = _cross_corr(self.submasks[i], self.tmpimgs[i] *
+                                         self.submasks[i])
+                else:
+                    Icorr2 = _cross_corr(self.submasks[i], self.tmpimgs2[i] *
+                                         self.submasks[i])
+                ccorr[w] *= self.maskcorrs[i][w]/Icorr[w]/Icorr2[w]
+
+            if 'regular' in normalization:
+                ccorr[w] /= self.maskcorrs[i][w] * \
+                            np.average(self.tmpimgs[i].
+                                       ravel()[self.subpxlsts[i]])**2
+
+            ccorrs.append(ccorr)
+
+        if len(ccorrs) == 1:
+            ccorrs = ccorrs[0]
+
+        return ccorrs
+
+
+def _cross_corr(img1, img2=None, axes=None):
+    ''' Compute the cross correlation of one (or two) images.
+
+        Parameters
+        ----------
+        img1 : 1d or 2d np.ndarray
+            the image or curve to cross correlate
+
+        img2 : 1d or 2d np.ndarray, optional
+            If set, cross correlate img1 against img2.  A shift of img2
+            to the right of img1 will lead to a shift of the point of
+            highest correlation to the right.
+            Default is set to None
+
+        axes : 1 or 2-tuple
+            the axis or axes to perform the cross correlation on
+    '''
+    if axes is None:
+        if img1.ndim == 2:
+            axes = -2, -1
+        else:
+            axes = -1,
+
+    if img2 is None:
+        img2 = img1
+
+    imgc = fftshift(ifft2(fft2(img1, axes=axes) *
+                    np.conj(fft2(img2, axes=axes)), axes=axes).real)
```

Comment:
I vote using fftconvol. For the record, computation as of now is:

`ret = (np.fft.irfftn(sp1 * sp2, fshape)[fslice].copy())`

which is FFT^-1(FFT(A)*FFT(B))

but where the data set has been reshaped to a size more efficient for FFT (speed of calculation is highly dependent upon array size being a multiple of some magic numbers). Makes more sense to use. (there's an if statement to handle complex arrays as well)